### PR TITLE
[WIP] Add preliminary support for libc & dynamic-linking

### PR DESCRIPTION
**Pull request by mwilliammyers**
Despite it adding the appropriate flags to the command, I can't seem to get this to work when compiling python 2.7.11 on our cluster even though I got it working manually (without spack) with a very similar method. Can anyone else get it to work./know what might be the issue?

```
...
checking build system type... x86_64-unknown-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking for --enable-universalsdk... no
checking for --with-universal-archs... 32-bit
checking MACHDEP... linux2
checking EXTRAPLATDIR... 
checking for --without-gcc... no
checking for gcc... $SPACK_ROOT/lib/spack/env/gcc/gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... configure: error: in `/tmp/$USER/spack-stage/spack-stage-zA2Itz/Python-2.7.11':
configure: error: cannot run C compiled programs.
If you meant to cross compile, use `--host'.
See `config.log' for more details
```

and here is the debug dump from the `cc` compiler wrapper:

```
[ccld] /apps/gcc/5.3.0/bin/gcc -Wl,-rpath=$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/glibc-2.17-trpcxsgryqgtsqbp5mvspmibcxyjfg57/lib -Wl,--dynamic-linker=$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/glibc-2.17-trpcxsgryqgtsqbp5mvspmibcxyjfg57/lib/ld-linux-x86-64.so.2 -nostdinc -L$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/glibc-2.17-trpcxsgryqgtsqbp5mvspmibcxyjfg57/lib -L/apps/gcc/5.3.0/lib64 -Wl,-rpath,/apps/gcc/5.3.0/lib64 -Wl,-rpath,$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/python-2.7.11-6usjtxyhy7qecbewxfmjnjeguqffrpme/lib -Wl,-rpath,$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/python-2.7.11-6usjtxyhy7qecbewxfmjnjeguqffrpme/lib64 -L$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/sqlite-3.8.5-qspfldeculrro7tekx7g4ezxrchf6pl7/lib -Wl,-rpath,$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/sqlite-3.8.5-qspfldeculrro7tekx7g4ezxrchf6pl7/lib -I$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/sqlite-3.8.5-qspfldeculrro7tekx7g4ezxrchf6pl7/include -L$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/readline-6.3-voguswurxi45eq7siehym4uxhsgfot3u/lib -Wl,-rpath,$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/readline-6.3-voguswurxi45eq7siehym4uxhsgfot3u/lib -I$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/readline-6.3-voguswurxi45eq7siehym4uxhsgfot3u/include -L$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/zlib-1.2.8-4smfllshblurbkwmozx7l6sgyz2nmtth/lib -Wl,-rpath,$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/zlib-1.2.8-4smfllshblurbkwmozx7l6sgyz2nmtth/lib -I$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/zlib-1.2.8-4smfllshblurbkwmozx7l6sgyz2nmtth/include -L$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/openssl-1.0.2h-zz2clu4raqeao4sy62a65w3xvigyfpxp/lib -Wl,-rpath,$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/openssl-1.0.2h-zz2clu4raqeao4sy62a65w3xvigyfpxp/lib -I$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/openssl-1.0.2h-zz2clu4raqeao4sy62a65w3xvigyfpxp/include -L$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/ncurses-6.0-avvusf6rjwxrgvqmkcyxwox4vqj5glzc/lib -Wl,-rpath,$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/ncurses-6.0-avvusf6rjwxrgvqmkcyxwox4vqj5glzc/lib -I$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/ncurses-6.0-avvusf6rjwxrgvqmkcyxwox4vqj5glzc/include -L/usr/lib64 -Wl,-rpath,/usr/lib64 -L/usr/lib -Wl,-rpath,/usr/lib -I/usr/include -L$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/bzip2-1.0.6-3njle5iq35prjqqyj7yxxs3pxbhn3d4r/lib -Wl,-rpath,$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/bzip2-1.0.6-3njle5iq35prjqqyj7yxxs3pxbhn3d4r/lib -I$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/bzip2-1.0.6-3njle5iq35prjqqyj7yxxs3pxbhn3d4r/include -o conftest -I$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/openssl-1.0.2h-zz2clu4raqeao4sy62a65w3xvigyfpxp/include -I$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/bzip2-1.0.6-3njle5iq35prjqqyj7yxxs3pxbhn3d4r/include -I$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/readline-6.3-voguswurxi45eq7siehym4uxhsgfot3u/include -I$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/ncurses-6.0-avvusf6rjwxrgvqmkcyxwox4vqj5glzc/include -I$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/sqlite-3.8.5-qspfldeculrro7tekx7g4ezxrchf6pl7/include -I$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/zlib-1.2.8-4smfllshblurbkwmozx7l6sgyz2nmtth/include -L$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/openssl-1.0.2h-zz2clu4raqeao4sy62a65w3xvigyfpxp/lib -L$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/bzip2-1.0.6-3njle5iq35prjqqyj7yxxs3pxbhn3d4r/lib -L$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/readline-6.3-voguswurxi45eq7siehym4uxhsgfot3u/lib -L$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/ncurses-6.0-avvusf6rjwxrgvqmkcyxwox4vqj5glzc/lib -L$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/sqlite-3.8.5-qspfldeculrro7tekx7g4ezxrchf6pl7/lib -L$SPACK_ROOT/opt/spack/linux-redhat6-x86_64/gcc-5.3.0/zlib-1.2.8-4smfllshblurbkwmozx7l6sgyz2nmtth/lib conftest.c
```

Also this PR only attempts to add support for ELF... I would have to read up more on how macOS does it... I did try to guard against setting the dynamic-linker on macOS but it probably needs a little more on that front.

### CODE REVIEW

File: lib/spack/env/cc [hunk 1]
```diff
@@ -301,6 +306,18 @@ case "$mode" in
         args=("${args[@]}" ${SPACK_LDLIBS[@]}) ;;
 esac
 
+# Prepend SPACK_DYNAMIC_LINKER & add it's directory to rpath
+if [[ -n "$SPACK_DYNAMIC_LINKER" && -e "$SPACK_DYNAMIC_LINKER" ]]; then
+    dynamic_linker_dir="$(dirname $SPACK_DYNAMIC_LINKER)"
+    if [[ $mode == ccld ]]; then
+        $add_rpaths && args=("$dynamic_linker$SPACK_DYNAMIC_LINKER" "${args[@]}")
+        $add_rpaths && args=("-Wl,-rpath=$dynamic_linker_dir" "${args[@]}")
```
Comment (mwilliammyers) [inline lib/spack/env/cc]: Note how the `-Wl,-rpath=` syntax is slightly different from the normal `-Wl,-rpath,`, so I couldn't use `$rpath`. I am wondering if we should export yet another variable like `$SPACK_DYNAMIC_LINKER_RPATH` or something similar.

File: lib/spack/env/cc [hunk 2]
```diff
@@ -301,6 +306,18 @@ case "$mode" in
...
+    elif [[ $mode == ld ]]; then
```
Comment (mwilliammyers) [inline lib/spack/env/cc]: Does anyone know the correct syntax for `ld`? Documentation on this stuff is surprisingly scarce...

File: lib/spack/spack/build_environment.py [hunk 3]
```diff
@@ -250,8 +254,17 @@ def set_build_environment_variables(pkg, env):
         env.prepend_path('PATH', item)
     env.set_path(SPACK_ENV_PATH, env_paths)
 
+    dep_prefixes = []
+    for d in pkg.spec.traverse(root=False):
+        # Set the dynamic linker to the path specified by the dependency
+        dep_pkg = d.package_class
+        if dep_pkg.dynamic_linker:
```
Comment (mwilliammyers) [inline lib/spack/spack/build_environment.py]: Should we fail if something depends on two packages that provide the `dynamic_linker` property? Or should we silently accept only the latest one?

File: lib/spack/env/cc (hunk 1 above)
Comment (adamjstewart) [inline lib/spack/env/cc]: What's the difference?

File: lib/spack/env/cc (hunk 2 above)
Comment (mplegendre) [inline lib/spack/env/cc]: Are you seeing some behavior implying that you don't have correct ld syntax?  This looks correct to me.

Comment (mwilliammyers) [inline lib/spack/env/cc]: No, I just thought it might be slightly different because it is slightly different for `ccld`

**Comment by alalazo** [comment]
@mwilliammyers Any news on this? Is it still a work in progress?

**Comment by mwilliammyers** [comment]
@alalazo this semester school has gotten the best of me... I might have some more time soon to take a look at it again.

**Comment by alalazo** [comment]
@mwilliammyers I am closing this as it has been inactive for a while. Feel free to reopen if you get to it.

**Pull request closed without merging**